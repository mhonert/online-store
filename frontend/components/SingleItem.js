import React from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import DisplayError from './ErrorMessage';
import styled from 'styled-components';
import Head from 'next/head';

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }

  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;

export const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      largeImage
    }
  }
`;

const SingleItem = ({ id }) => {
  return (
    <Query query={SINGLE_ITEM_QUERY} variables={{ id: id }}>
      {({ error, data, loading }) => {
        if (error) return <DisplayError error={error} />;
        if (loading) return <p>Loading ...</p>;
        const item = data.item;
        if (!item) return <p>No item found for ID {id}</p>;

        return (
          <SingleItemStyles>
            <Head>
              <title>Fresh Fruits | {item.title}</title>
            </Head>
            <img src={item.largeImage} alt={item.title} />
            <div className="details">
              <h2>Viewing {item.title}</h2>
              <p>{item.description}</p>
            </div>
          </SingleItemStyles>
        );
      }}
    </Query>
  );
};

export default SingleItem;
