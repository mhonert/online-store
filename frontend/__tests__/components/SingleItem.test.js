import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import SingleItem, { SINGLE_ITEM_QUERY } from '../../components/SingleItem';
import { MockedProvider } from 'react-apollo/test-utils';

import { fakeItem } from '../../lib/testUtils';

describe('<SingleItem/>', () => {
  it('renders correctly', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        result: {
          data: {
            item: fakeItem()
          }
        }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );
    expect(wrapper.text()).toContain('Loading ...');
    await wait();
    wrapper.update();

    expect(toJSON(wrapper.find('SingleItem__SingleItemStyles'))).toMatchSnapshot();
  });

  it('errors with not found item', async() => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        result: {
          errors: [{message: 'Items Not Found!'}]
        }
      }
    ]
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('[data-test="graphql-error"]'))).toMatchSnapshot();
  })
});
