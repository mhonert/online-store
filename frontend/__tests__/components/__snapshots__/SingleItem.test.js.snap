// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem/> errors with not found item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Items Not Found!
</p>
`;

exports[`<SingleItem/> renders correctly 1`] = `
<SingleItem__SingleItemStyles>
  <div
    className="SingleItem__SingleItemStyles-sc-1u71o50-0 dbTpnF"
  >
    <SideEffect(Head)>
      <Head />
    </SideEffect(Head)>
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      className="details"
    >
      <h2>
        Viewing 
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </div>
</SingleItem__SingleItemStyles>
`;
